iree/vm/test/emitc/module_test.cc

```diff
@@ -0,0 +1,142 @@
+// Copyright 2021 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#include "absl/strings/match.h"
+#include "absl/strings/str_replace.h"
+#include "iree/base/logging.h"
+#include "iree/base/status.h"
+#include "iree/testing/gtest.h"
+#include "iree/vm/api.h"
+
+// This would be generated together with the functions in the header
+#include "iree/vm/test/emitc/arithmetic_ops.h"
+
+namespace {
+
+typedef iree_status_t (*create_function_t)(iree_allocator_t,
+                                           iree_vm_module_t**);
+
+struct TestParams {
+  std::string module_name;
+  std::string local_name;
+  create_function_t create_function;
+};
+
+struct ModuleDescription {
+  iree_vm_native_module_descriptor_t descriptor;
+  create_function_t create_function;
+};
+
+std::ostream& operator<<(std::ostream& os, const TestParams& params) {
+  return os << absl::StrReplaceAll(params.local_name, {{":", "_"}, {".", "_"}});
+}
+
+std::vector<TestParams> GetModuleTestParams() {
+  std::vector<TestParams> test_params;
+
+  // TODO(simon-camp) get these automatically
+  std::vector<ModuleDescription> modules = {
```

Comment:
It'd be cool to share this entire file with both emitc and the bytecode; a `iree/vm/testing/module_test.h+cc` that contained most of this and then an extern "GetModuleFactory()" that would get defined by the particular test type (`iree/vm/bytecode_module_test_factory.cc`, etc). Maybe file an issue so that we can track that? (it'd also be useful for benchmarks and such too)